temporal-sdk/src/test/java/io/temporal/worker/CleanWorkerShutdownHeartBeatingActivityTest.java

```diff
@@ -0,0 +1,122 @@
+import io.temporal.testing.internal.SDKTestOptions;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.shared.TestActivities.NoArgsReturnsStringActivity;
+import io.temporal.workflow.shared.TestWorkflows.TestWorkflowReturnString;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class CleanWorkerShutdownHeartBeatingActivityTest {
+
+  private static final String EXPECTED_RESULT = "Worker Shutdown";
+  private static final CompletableFuture<Boolean> started = new CompletableFuture<>();
+  private static final HeartBeatingActivitiesImpl activitiesImpl =
+      new HeartBeatingActivitiesImpl(started);
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkflowTypes(TestWorkflowImpl.class)
+          .setActivityImplementations(activitiesImpl)
+          .build();
+
+  /**
+   * Tests that Activity#heartbeat throws ActivityWorkerShutdownException after {@link
+   * WorkerFactory#shutdown()} is closed.
+   */
+  @Test
+  public void testShutdownHeartBeatingActivity() throws ExecutionException, InterruptedException {
+    TestWorkflowReturnString workflow =
+        testWorkflowRule.newWorkflowStub(TestWorkflowReturnString.class);
+    WorkflowExecution execution = WorkflowClient.start(workflow::execute);
+    started.get();
+    testWorkflowRule.getTestEnvironment().shutdown();
+    testWorkflowRule.getTestEnvironment().awaitTermination(10, TimeUnit.MINUTES);
+    List<HistoryEvent> events = testWorkflowRule.getHistory(execution).getEventsList();
+    boolean found = false;
+    for (HistoryEvent e : events) {
+      if (e.getEventType() == EventType.EVENT_TYPE_ACTIVITY_TASK_COMPLETED) {
+        found = true;
+        Payloads ar = e.getActivityTaskCompletedEventAttributes().getResult();
+        String r =
+            DataConverter.getDefaultInstance()
+                .fromPayloads(0, Optional.of(ar), String.class, String.class);
+        assertEquals(EXPECTED_RESULT, r);
+      }
+    }
+    assertTrue("Contains ActivityTaskCompleted", found);
+  }
+
+  public static class TestWorkflowImpl implements TestWorkflowReturnString {
+
+    private final NoArgsReturnsStringActivity activities =
+        Workflow.newActivityStub(
+            NoArgsReturnsStringActivity.class,
+            SDKTestOptions.newActivityOptions20sScheduleToClose());
+
+    @Override
+    public String execute() {
+      return activities.execute();
+    }
+  }
+
+  public static class HeartBeatingActivitiesImpl implements NoArgsReturnsStringActivity {
+    private final CompletableFuture<Boolean> started;
+
+    public HeartBeatingActivitiesImpl(CompletableFuture<Boolean> started) {
+      this.started = started;
+    }
+
+    @Override
```

Comment:
This doesn't look right and I'm not sure if it's properly testing what is intended to be tested. It looks like the right implementation will be something like...


    @Override
     public String execute() {
         started.complete(true);

       try {
         Thread.sleep(1000);
       } catch (InterruptedException e) {
            //we ignore interrupted exception here, because the main goal of `testShutdownHeartBeatingActivity` test is to check that heartbeat returns a ActivityWorkerShutdownException
       }
       try {
           Activity.getExecutionContext().heartbeat("foo");
       } catch (ActivityWorkerShutdownException e) {
         return EXPECTED_RESULT;
       }

       return "completed"; //should never happen
     }